### DISCUSSION TITLE
Is there a way to validate an archive for the App Store with deliver?

### DISCUSSION THREAD

**Discussion by polpielladev** [fastlane/fastlane#20204] [General] [answered]
In my team, we're currently looking at migrating our release process from using a combination of `altool` and App Store Connect username/password to using `deliver` with the App Store Connect API.

The upload to App Store Connect seems very straightforward but the issue is that we have a nightly lane that validates an archive and sees if it is suitable for submission to the App Store by using `altool`:

```bash
xcrun altool --validate-app -f ipa_path -t ios -u username -p password
```

I am aware that we can still use `altool` and pass in the contents of the App Store Connect api key, but we would ideally want to sunset `altool` altogether and replace it with a fastlane action. Is there anything we can use? Or could this be a good idea to add a `validate` option to `deliver` and see if this logic can be done there? I would not mind looking into it if people think it's a good idea.

👉 [Related Issue from a while back](https://github.com/fastlane/fastlane/issues/15441)

**Comment by lucgrabowski** [comment, marked as answer]
Hi @pol-piella I currently don't see a way to validate a build with fastlane/deliver. 
I see two ways of handling this: 

1. Use of `altool` could be wrapped in fastlane action or used in a custom plugin https://docs.fastlane.tools/plugins/create-plugin/

2. Using transporter to verify the build, similarly to uploading a build. 
As you say, with `validate` option added to `deliver`, `deliver` could verify the build instead of uploading it to appstore connect.
I've briefly taken a look at transporter guide https://help.apple.com/itc/transporteruserguide/en.lproj/static.html#apdATD1E2248-D1E1A1303-D1E2248A1126 and `verify` mode looks promising.

If you want to see how build's are uploaded with transporter, you could take a look at `build_upload_command`s in https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/itunes_transporter.rb. I suppose that spec could also be helpful in understanding the code: https://github.com/fastlane/fastlane/blob/master/fastlane_core/spec/itunes_transporter_spec.rb

Transporter is used in `deliver`'s `Runner`:
https://github.com/fastlane/fastlane/blob/90a034505d8fc944184fb8e1db24cd44c88c16c1/deliver/lib/deliver/runner.rb#L159-L195
https://github.com/fastlane/fastlane/blob/90a034505d8fc944184fb8e1db24cd44c88c16c1/deliver/lib/deliver/runner.rb#L232-L255


What do you think about it?

**Reply by polpielladev** [reply]
Hey @lucgrabowski 👋  Thanks for the super quick reply, I really appreciate it!

I am a big fan of the second idea, I might have a go at implementing it and submit a PR for it to see if people are happy to get it in? Is there a specific process I need to follow?

Thanks 🙌

**Reply by lucgrabowski** [reply]
Hi @pol-piella you're welcome. Have fun with implementation :)
About the process:
There is a page with https://github.com/fastlane/fastlane/blob/master/YourFirstPR.md that might be helpful.
And when creating a PR you will see checkboxes with things that should be done:
1. I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
2. I've followed the fastlane code style and run `bundle exec rubocop -a` to ensure the code style is valid
3. I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
4. I've updated the documentation if necessary.

**Reply by polpielladev** [reply]
Fantastic! I'll get something together 👍 

Thanks so much @lucgrabowski 🎉

**Comment by polpielladev** [comment]
Hi @lucgrabowski 👋 

I have opened a Pull Request to for these changes to `deliver`, would you be able to take a look and see if they make sense to you when you get a chance?

👉 PR: https://github.com/fastlane/fastlane/pull/20247

Also, I have a couple of questions which I am not too sure how to tackle:

1. I have a couple of test failures that seem unrelated to my changes locally, but not sure how to fix them or if they are a red herring?
<img width="1436" alt="CleanShot 2022-05-06 at 11 14 45@2x" src="https://user-images.githubusercontent.com/24246926/167112972-841574f3-1feb-45ae-a4d3-a9ca835cf6ac.png">

2. How would I update the docs to reflect the changes on the PR? Is this something that gets automatically done by a docs generator based on the flags?

**Reply by lucgrabowski** [reply]
Thanks for submitting PR with implementation and with tests :) I'll try to review it as soon as possible. I've just taken a look at it briefly to be able to respond to quesions:
1. I pulled the code and ran test suite and all the tests are running successfully for me - the same way as you can see in CI checks in PR.
2. Docs are done by a generator, docs.fastlane.tools/actions/deliver/#parameters is updated based on `FastlaneCore::ConfigItem`'s description. It will be done after PR is merged.
